mne/time_frequency/tfr.py

```diff
@@ -616,35 +746,101 @@ def _tfr_aux(method, inst, freqs, decim, return_itc, picks, average,
             raise ValueError('Inter-trial coherence is not supported'
                              ' with average=False')
 
-    out = _compute_tfr(data, freqs, info['sfreq'], method=method,
-                       output=output, decim=decim, **tfr_params)
+    info = None
+    if isinstance(inst, list) or isgenerator(inst):
+
+        out, inst = _tfr_loop_list(inst, freqs, method=method,
+                                   output=output, decim=decim, **tfr_params)
+        # nave is not needed for any SourceTFR type
+        nave = None
+    else:
+        data = _get_data(inst, return_itc)
+        if average:
+            nave = len(data)
+        if isinstance(inst, _BaseSourceEstimate):
+            sfreq = inst.sfreq
+        else:
+            info, data = _prepare_picks(inst.info, data, picks, axis=1)
+            sfreq = inst.info['sfreq']
+            del picks
+
+        out = _compute_tfr(data, freqs, sfreq, method=method,
+                           output=output, decim=decim, **tfr_params)
+
+    if average and return_itc:
+        power, itc = out.real, out.imag
+    else:
+        power = out
+
     times = inst.times[decim].copy()
 
+    # put the output objects together accordingly
     if average:
+        out = assign_tfr_class(power, inst, info, freqs, times, average,
+                               method='{}-power'.format(method), nave=nave)
         if return_itc:
-            power, itc = out.real, out.imag
-        else:
-            power = out
-        nave = len(data)
-        out = AverageTFR(info, power, times, freqs, nave,
-                         method='%s-power' % method)
-        if return_itc:
-            out = (out, AverageTFR(info, itc, times, freqs, nave,
-                                   method='%s-itc' % method))
+            out = (out, assign_tfr_class(itc, inst, info, freqs, times,
+                                         average, nave=nave,
+                                         method='{}-itc'.format(method)))
+    else:
+        out = assign_tfr_class(power, inst, info, freqs, times, average,
+                               method='{}-power'.format(method))
```

Comment:
@massich
I tried to refactor this now (as you noted in https://github.com/mne-tools/mne-python/pull/6629#discussion_r310627677 ):
Some parts still look ugly to me, especially the assignment of `info` and `nave` (lines 749 and 755). What do you say?
Is there maybe some elegant way we can improve this? Or should I maybe try a different attempt for the refactoring?